astropy/units/format/base.py

```diff
@@ -133,37 +122,49 @@ def to_string(
         ValueError
             If ``fraction`` is not recognized.
         """
+        string_components = cls._to_string_helper(unit, fraction)
+        if isinstance(string_components, str):
+            return string_components
+        if fraction is True or fraction == "inline":
+            return cls._format_inline_fraction(*string_components)
+        raise ValueError(
+            f"format {cls.name!r} only supports 'inline' fractions, not {fraction=!r}."
+        )
+
+    @classmethod
+    def _to_string_helper(
+        cls, unit: UnitBase, fraction: bool | Literal["inline", "multiline"]
+    ) -> str | tuple[str, str, str]:
+        # The `unit` is converted to a string if there is no need to decide how
+        # to handle fractions. Otherwise the scale, numerator and denominator
+        # get returned so that `to_string()` could decide how to format them.
+
         # First the scale.  Normally unity, in which case we omit
         # it, but non-unity scale can happen, e.g., in decompositions
         # like u.Ry.decompose(), which gives "2.17987e-18 kg m2 / s2".
-        if unit.scale == 1:
-            s = ""
-        else:
-            s = cls.format_exponential_notation(unit.scale)
-
-        # Now the unit baes, taking care that dimensionless does not have any
-        # (but can have a scale; e.g., u.percent.decompose() gives "0.01").
-        if len(unit.bases):
-            if s:
-                s += cls._scale_unit_separator
-            if fraction:
-                numerator, denominator = utils.get_grouped_by_powers(
-                    unit.bases, unit.powers
-                )
-            else:
-                numerator = list(zip(unit.bases, unit.powers))
-                denominator = []
-            if len(denominator):
-                if len(numerator):
-                    numerator = cls._format_unit_list(numerator)
-                else:
-                    numerator = "1"
-                denominator = cls._format_unit_list(denominator)
-                s = cls._format_fraction(s, numerator, denominator, fraction=fraction)
+        s = "" if unit.scale == 1 else cls.format_exponential_notation(unit.scale)
+
+        # dimensionless does not have any bases, but can have a scale;
+        # e.g., u.percent.decompose() gives "0.01".
+        if not unit.bases:
+            return s
+
+        if s:
+            s += cls._scale_unit_separator
+        if not fraction:
+            return s + cls._format_unit_list(zip(unit.bases, unit.powers))
+        positive = []
+        negative = []
+        for base, power in zip(unit.bases, unit.powers):
```

Comment:
?
```suggestion
        for base, power in zip(unit.bases, unit.powers, strict=True):
```

Reply:
What we should be discussing instead is why are `bases` and `powers` stored two separate `list` instances instead of being stored in a single `dict`. If we could call `unit.bases_and_powers.items()` then we wouldn't have to discuss the `strict` argument at all.

Reply:
Probably comes all the way from `pynbody`...

Reply:
That would be a nice refactor.

Reply:
I found out why the bases and powers are not keys and values in a `dict`. It's because using units as keys requires hashing them, but `UnitBase.__hash__()` requires bases and powers to already be defined. This is a problem because sometimes a unit contains itself in its bases.

Storing the components of a unit as `dict[UnitBase, Real]` wouldn't not work, but storing them as `tuple[tuple[UnitBase, Real], ...]` might still be a good idea. In that case we could replace all calls of `for b, p in zip(unit.bases, unit.powers)` with `for b, p in unit.components`, which is shorter even if we don't add `strict=True` to the `zip()` call. It would also be very simple to re-implement `bases` and `powers` to maintain backwards compatibility. Is this idea worth pursuing?

Reply:
Ah, indeed, that makes sense (and is presumably why `base.name` is used in the hash).

My sense is that the tuple of 2-tuple idea is not quite enough of a benefit to pursue, especially since the signature of `CompositeUnit` uses the bases and powers structure.

Reply:
> This is a problem because sometimes a unit contains itself in its bases.

That seems strange! When does this occur? And can we refactor to avoid this?

Reply:
It's not actually strange:
```python
>>> from astropy import units as u
>>> u.m.bases
[Unit("m")]
>>> u.m.bases[0] is u.m
True
```

Reply:
Definitely not for this PR, but I've never loved that base units are aware of how they decompose relative to other units. IMO we shouldn't be encountering this. Units should hold dimensions and powers, e.g. https://pint.readthedocs.io/en/stable/api/base.html#pint.Unit.dimensionality

Reply:
I don't think it is that bad - it just means that one can treat base units, prefix units, and composite units all in the same way.  Anyway, indeed definitely out of scope here!